File: leetcode/0001-1000/0605-can-place-flowers/CanPlaceFlowers.java
public class CanPlaceFlowers {
  public static void main(String[] args) {
    final int[] flowerbed = { 1, 0, 0, 0, 1, 0, 0 };
    final int n = 2;

    System.out.println((new Solution()).canPlaceFlowers(flowerbed, n));
  }
}

class Solution {
  public boolean canPlaceFlowers(int[] flowerbed, int n) {
    int extra = 0;
    int length = flowerbed.length;

    for (int i = 0; i < length && extra < n; i++) {
      int left = i == 0 ? 0 : flowerbed[i - 1];
      int right = i == length - 1 ? 0 : flowerbed[i + 1];

      if (right + left + flowerbed[i] == 0) {
        flowerbed[i] = 1;

        extra++;
      }
    }

    return extra >= n;
  }
}

File: leetcode/0001-1000/0605-can-place-flowers/CanPlaceFlowers.ts
function canPlaceFlowers(flowerbed: number[], n: number): boolean {
  const length = flowerbed.length;

  const extra = flowerbed.reduce(
    (extra, currentValue, currentIndex, bed) => {
      const left = currentIndex === 0 ? 0 : bed[currentIndex - 1];
      const right = currentIndex === length - 1 ? 0 : bed[currentIndex + 1];

      if (left + right + currentValue === 0) {
        bed[currentIndex] = 1;

        return extra + 1;
      }

      return extra;
    }, 0);

  return extra >= n;
};

console.log(canPlaceFlowers([1, 0, 0, 0, 1, 0, 0], 2));
